ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/validation/ErrorFactories.scala

```diff
@@ -4,54 +4,205 @@
-    grpcError(Status.UNAUTHENTICATED)
+    grpcError(
+      Status
+        .newBuilder()
+        .setCode(Code.UNAUTHENTICATED.value())
+        .build()
+    )
 
-  def missingLedgerConfig(): StatusRuntimeException =
-    grpcError(Status.UNAVAILABLE.withDescription("The ledger configuration is not available."))
+  /** @param definiteAnswer A flag that says whether it is a definite answer. Provided only in the context of command deduplication.
+    * @return An exception with the [[Code.UNAVAILABLE]] status code.
+    */
+  def missingLedgerConfig(definiteAnswer: Option[Boolean]): StatusRuntimeException = {
+    val statusBuilder = Status
+      .newBuilder()
+      .setCode(Code.UNAVAILABLE.value())
+      .setMessage("The ledger configuration is not available.")
+    addDefiniteAnswerDetails(definiteAnswer, statusBuilder)
+    grpcError(statusBuilder.build())
+  }
 
   def missingLedgerConfigUponRequest(): StatusRuntimeException =
-    grpcError(Status.NOT_FOUND.withDescription("The ledger configuration is not available."))
+    grpcError(
+      Status
+        .newBuilder()
+        .setCode(Code.NOT_FOUND.value())
+        .setMessage("The ledger configuration is not available.")
+        .build()
+    )
 
   def participantPrunedDataAccessed(message: String): StatusRuntimeException =
-    grpcError(Status.NOT_FOUND.withDescription(message))
-
-  def grpcError(status: Status): StatusRuntimeException =
-    new ApiException(status)
+    grpcError(
+      Status
+        .newBuilder()
+        .setCode(Code.NOT_FOUND.value())
+        .setMessage(message)
+        .build()
+    )
 
+  /** @param definiteAnswer A flag that says whether it is a definite answer. Provided only in the context of command deduplication.
+    * @return An exception with the [[Code.UNAVAILABLE]] status code.
+    */
+  def serviceNotRunning(definiteAnswer: Option[Boolean]): StatusRuntimeException = {
+    val statusBuilder = Status
+      .newBuilder()
+      .setCode(Code.UNAVAILABLE.value())
+      .setMessage("Service has been shut down.")
+    addDefiniteAnswerDetails(definiteAnswer, statusBuilder)
+    grpcError(statusBuilder.build())
+  }
+
+  def grpcError(status: Status): StatusRuntimeException = new ApiException(
+    StatusProto.toStatusRuntimeException(status)
+  )
 }
 
-object ErrorFactories extends ErrorFactories
+object ErrorFactories extends ErrorFactories {
+  private[daml] lazy val definiteAnswers = Map(
+    true -> AnyProto.pack[ErrorInfo](
+      ErrorInfo.newBuilder().putMetadata(GrpcStatuses.DefiniteAnswerKey, "true").build()
+    ),
+    false -> AnyProto.pack[ErrorInfo](
+      ErrorInfo.newBuilder().putMetadata(GrpcStatuses.DefiniteAnswerKey, "false").build()
+    ),
+  )
+
+  private def addDefiniteAnswerDetails(
+      definiteAnswer: Option[Boolean],
+      statusBuilder: Status.Builder,
+  ): Unit = {
+    definiteAnswer.foreach { definiteAnswer =>
+      statusBuilder.addDetails(definiteAnswers(definiteAnswer))
+    }
+  }
```

Comment:
I find this a lot more difficult to conceptualize in my mind, I think a simple if inside the foreach would make a lot more sense. This would allow us to also keep the FalseDefinitiveAnswer, TrueDefinitiveAnswer constants.

Reply:
There are multiple places where the map is being used, so it reduces duplication. I would leave it as is.